clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -7430,6 +7430,169 @@ static void handleSYCLIntelFPGAMaxConcurrencyAttr(Sema &S, Decl *D,
   S.AddSYCLIntelFPGAMaxConcurrencyAttr(D, A, E);
 }
 
+// Returns true if an error occured.
+static bool evaluateAddIRAttributesArgs(Expr **ArgsBegin, size_t ArgsSize,
+                                        Sema *S,
+                                        const AttributeCommonInfo &CI) {
+  ASTContext &Context = S->getASTContext();
+
+  // Check filter list if it is the first argument.
+  bool HasFilter = ArgsSize && isa<InitListExpr>(ArgsBegin[0]);
+  if (HasFilter) {
+    const auto *FilterListE = cast<InitListExpr>(ArgsBegin[0]);
+    for (const Expr *FilterElemE : FilterListE->inits()) {
+      if (!isa<StringLiteral>(FilterElemE)) {
+        S->Diag(FilterElemE->getBeginLoc(),
+                diag::err_sycl_add_ir_attribute_invalid_filter)
+            << CI;
+        return true;
+      }
+    }
+  }
+
+  llvm::SmallVector<PartialDiagnosticAt, 8> Notes;
+  bool HasDependentArg = false;
+  for (unsigned I = HasFilter; I < ArgsSize; I++) {
+    Expr *&E = ArgsBegin[I];
+    assert(E && "error are handled before");
+
+    bool Result = false;
+    if (isa<InitListExpr>(E)) {
+      S->Diag(E->getBeginLoc(), diag::err_add_ir_attr_filter_list_invalid_arg)
+          << CI;
+      return true;
+    }
+
+    if (E->isValueDependent() || E->isTypeDependent()) {
+      HasDependentArg = true;
+      continue;
+    }
+
+    Expr::EvalResult Eval;
+    Eval.Diag = &Notes;
+    Result = E->EvaluateAsConstantExpr(Eval, Context);
+    if (Result && Notes.empty()) {
```

Comment:
You can move this below the diagnostic check below. You won't need the `if `then.